### ISSUE TITLE
Can't spawn asynk::Subscription

### ISSUE THREAD

**Issue by andrewbanchich** [nats-io/nats.rs#92]
I'm using `smol` and `nats`, and just migrated from `rants` to this official client since it now has direct async support.

With `rants`, I was creating a subscription stream and mapping messages to their deserialized data, then spawning a task to run through that stream in the background. After migrating, I'm getting this:

```
error[E0277]: `std::sync::MutexGuard<'_, std::collections::HashSet<u64>>` cannot be sent between threads safely
   --> detect/src/main.rs:72:5
    |
72  |     Task::spawn(async move {
    |     ^^^^^^^^^^^ `std::sync::MutexGuard<'_, std::collections::HashSet<u64>>` cannot be sent between threads safely
    |
   ::: /Users/andrewbanchich/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/src/libcore/future/mod.rs:55:43
    |
55  | pub const fn from_generator<T>(gen: T) -> impl Future<Output = T::Return>
    |                                           ------------------------------- within this `impl core::future::future::Future`
    |
   ::: /Users/andrewbanchich/.cargo/registry/src/github.com-1ecc6299db9ec823/async-executor-0.1.2/src/lib.rs:421:31
    |
421 |     pub fn spawn(future: impl Future<Output = T> + Send + 'static) -> Task<T>
    |                               ------------------ required by this bound in `async_executor::Task::<T>::spawn`
```

**Comment by spacejam** [comment]
@andrewbanchich as a short-term workaround before we get the MutexGuard away from a yield point, you can spawn the future using `smol::Task::local` instead of `smol::Task::spawn` which uses the local thread, instead of a multi-threaded work stealing executor.

**Comment by andrewbanchich** [comment]
Thanks, I'll do that for now.

**Comment by stjepang** [comment]
Another option is to simply use a different mutex implementation where `MutexGuard` is `Send` and `Sync`. :) 

I can recommend two options:

* https://docs.rs/parking_lot/0.11.0/parking_lot/type.Mutex.html
* https://docs.rs/simple-mutex/1.1.5/simple_mutex/struct.Mutex.html

**Comment by sergeyzenchenko** [comment]
Hi guys, any updates on this issue? it kind of block async usage for many scenarios.

**Comment by stjepang** [comment]
@sergeyzenchenko  This issue is not really related to the `nats` crate and is resolved by using `parking_lot::Mutex` or `simple_mutex::Mutex` instead of `std::sync::Mutex`.

But, could you perhaps elaborate a little bit, share a piece of code that doesn't compile or the error message you're getting? That way I could be of more help.

**Comment by andrewbanchich** [comment]
@stjepang The error I posted wasn't from a Mutex I was using - it was one `nats` uses internally.

**Comment by andrewbanchich** [comment]
@stjepang I'm not using a `Mutex<HashSet<u64>>` so I think this is coming from a dependency I have no control over.

**Comment by stjepang** [comment]
Interesting - `nats` is not using `HashSet<u64>` either so now we just gotta track down the crate it's coming from.

Usually, when you get errors about `Send` and `Sync`, the compiler prints a long list of lines explaining why something was required to be `Send` or `Sync`, something like:

> A must be Send because used here
> B must be Sync because A appears in B and used here
> ...
> ...

Can you perhaps share this part of the compilation error message?

**Comment by andrewbanchich** [comment]
Sure, I'll look into this Monday and see what I can find.

**Comment by sergeyzenchenko** [comment]
Hi, sorry for late replay. I've tried to reproduce this issue in separate project and it works fine. Looks like something wrong with my code.

**Comment by sergeyzenchenko** [comment]
The error message is incredible verbose so it was not enable to find the root cause yet. It did partially mentioned parts of nats itself so I thought that it's something related to this issues since I am getting same error about `std::sync::MutexGuard<'_, std::collections::HashSet<u64>>` cannot be sent between threads safely`

**Comment by sergeyzenchenko** [comment]
Looks like I've been able to reproduce this issues. It happens when async_trair is used. You can find example here https://github.com/sergeyzenchenko/nats.rs-async-issue I am not sure is it nats issue or async_trait issue yet.

**Comment by sergeyzenchenko** [comment]
If you build this project you will get next error:

```
error[E0277]: `std::sync::MutexGuard<'_, std::collections::HashSet<u64>>` cannot be sent between threads safely
  --> src/main.rs:33:49
   |
33 |       async fn get(&self, data: &Data) -> Vec<u8> {
   |  _________________________________________-------_^
   | |                                         |
   | |                                         within this `impl std::future::Future`
34 | |         let result = self.nc.request("demo", data.name.as_bytes()).await.unwrap();
35 | |         result.data
36 | |     }
   | |_____^ `std::sync::MutexGuard<'_, std::collections::HashSet<u64>>` cannot be sent between threads safely
   |
   = help: within `impl std::future::Future`, the trait `std::marker::Send` is not implemented for `std::sync::MutexGuard<'_, std::collections::HashSet<u64>>`
   = note: required because it appears within the type `for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10, 't11, 't12, 't13, 't14, 't15, 't16, 't17, 't18> {std::future::ResumeTy, &'r nats::asynk::client::Client, &'s mut async_mutex::MutexGuard<'t0, nats::asynk::client::State>, nats::asynk::client::Client, &'t1 std::sync::Mutex<std::collections::HashSet<u64>>, std::sync::Arc<std::sync::Mutex<std::collections::HashSet<u64>>>, std::result::Result<std::sync::MutexGuard<'t2, std::collections::HashSet<u64>>, std::sync::PoisonError<std::sync::MutexGuard<'t3, std::collections::HashSet<u64>>>>, &'t4 mut std::sync::MutexGuard<'t5, std::collections::HashSet<u64>>, std::sync::MutexGuard<'t6, std::collections::HashSet<u64>>, std::collections::HashSet<u64>, &'t7 mut std::collections::HashSet<u64>, std::collections::hash_set::IntoIter<u64>, u64, nats::asynk::client::State, &'t8 mut std::option::Option<std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't9)>>>, std::option::Option<std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't10)>>>, std::option::Option<&'t11 mut std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't12)>>>, &'t13 mut std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't14)>>, std::option::Option<u64>, nats::asynk::proto::ClientOp<'t15>, impl std::future::Future, ()}`
   = note: required because it appears within the type `[static generator@nats::asynk::client::Client::cleanup_subscriptions::{{closure}}#0 0:&nats::asynk::client::Client, 1:&mut async_mutex::MutexGuard<'_, nats::asynk::client::State> for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10, 't11, 't12, 't13, 't14, 't15, 't16, 't17, 't18> {std::future::ResumeTy, &'r nats::asynk::client::Client, &'s mut async_mutex::MutexGuard<'t0, nats::asynk::client::State>, nats::asynk::client::Client, &'t1 std::sync::Mutex<std::collections::HashSet<u64>>, std::sync::Arc<std::sync::Mutex<std::collections::HashSet<u64>>>, std::result::Result<std::sync::MutexGuard<'t2, std::collections::HashSet<u64>>, std::sync::PoisonError<std::sync::MutexGuard<'t3, std::collections::HashSet<u64>>>>, &'t4 mut std::sync::MutexGuard<'t5, std::collections::HashSet<u64>>, std::sync::MutexGuard<'t6, std::collections::HashSet<u64>>, std::collections::HashSet<u64>, &'t7 mut std::collections::HashSet<u64>, std::collections::hash_set::IntoIter<u64>, u64, nats::asynk::client::State, &'t8 mut std::option::Option<std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't9)>>>, std::option::Option<std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't10)>>>, std::option::Option<&'t11 mut std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't12)>>>, &'t13 mut std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't14)>>, std::option::Option<u64>, nats::asynk::proto::ClientOp<'t15>, impl std::future::Future, ()}]`
   = note: required because it appears within the type `std::future::from_generator::GenFuture<[static generator@nats::asynk::client::Client::cleanup_subscriptions::{{closure}}#0 0:&nats::asynk::client::Client, 1:&mut async_mutex::MutexGuard<'_, nats::asynk::client::State> for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10, 't11, 't12, 't13, 't14, 't15, 't16, 't17, 't18> {std::future::ResumeTy, &'r nats::asynk::client::Client, &'s mut async_mutex::MutexGuard<'t0, nats::asynk::client::State>, nats::asynk::client::Client, &'t1 std::sync::Mutex<std::collections::HashSet<u64>>, std::sync::Arc<std::sync::Mutex<std::collections::HashSet<u64>>>, std::result::Result<std::sync::MutexGuard<'t2, std::collections::HashSet<u64>>, std::sync::PoisonError<std::sync::MutexGuard<'t3, std::collections::HashSet<u64>>>>, &'t4 mut std::sync::MutexGuard<'t5, std::collections::HashSet<u64>>, std::sync::MutexGuard<'t6, std::collections::HashSet<u64>>, std::collections::HashSet<u64>, &'t7 mut std::collections::HashSet<u64>, std::collections::hash_set::IntoIter<u64>, u64, nats::asynk::client::State, &'t8 mut std::option::Option<std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't9)>>>, std::option::Option<std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't10)>>>, std::option::Option<&'t11 mut std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't12)>>>, &'t13 mut std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't14)>>, std::option::Option<u64>, nats::asynk::proto::ClientOp<'t15>, impl std::future::Future, ()}]>`
   = note: required because it appears within the type `impl std::future::Future`
   = note: required because it appears within the type `impl std::future::Future`
   = note: required because it appears within the type `for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10, 't11, 't12, 't13, 't14, 't15, 't16, 't17, 't18, 't19, 't20> {std::future::ResumeTy, &'r nats::asynk::client::Client, &'s str, std::option::Option<&'t0 str>, impl std::future::Future, (), nats::asynk::client::Client, &'t1 async_mutex::Mutex<nats::asynk::client::State>, std::sync::Arc<async_mutex::Mutex<nats::asynk::client::State>>, impl std::future::Future, async_mutex::MutexGuard<'t4, nats::asynk::client::State>, &'t5 mut async_mutex::MutexGuard<'t6, nats::asynk::client::State>, impl std::future::Future, u64, nats::asynk::client::State, &'t10 mut std::option::Option<std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't11)>>>, std::option::Option<std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't12)>>>, std::option::Option<&'t13 mut std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't14)>>>, &'t15 mut std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't16)>>, nats::asynk::proto::ClientOp<'t17>, impl std::future::Future}`
   = note: required because it appears within the type `[static generator@nats::asynk::client::Client::subscribe::{{closure}}#0 0:&nats::asynk::client::Client, 1:&str, 2:std::option::Option<&str> for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10, 't11, 't12, 't13, 't14, 't15, 't16, 't17, 't18, 't19, 't20> {std::future::ResumeTy, &'r nats::asynk::client::Client, &'s str, std::option::Option<&'t0 str>, impl std::future::Future, (), nats::asynk::client::Client, &'t1 async_mutex::Mutex<nats::asynk::client::State>, std::sync::Arc<async_mutex::Mutex<nats::asynk::client::State>>, impl std::future::Future, async_mutex::MutexGuard<'t4, nats::asynk::client::State>, &'t5 mut async_mutex::MutexGuard<'t6, nats::asynk::client::State>, impl std::future::Future, u64, nats::asynk::client::State, &'t10 mut std::option::Option<std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't11)>>>, std::option::Option<std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't12)>>>, std::option::Option<&'t13 mut std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't14)>>>, &'t15 mut std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't16)>>, nats::asynk::proto::ClientOp<'t17>, impl std::future::Future}]`
   = note: required because it appears within the type `std::future::from_generator::GenFuture<[static generator@nats::asynk::client::Client::subscribe::{{closure}}#0 0:&nats::asynk::client::Client, 1:&str, 2:std::option::Option<&str> for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10, 't11, 't12, 't13, 't14, 't15, 't16, 't17, 't18, 't19, 't20> {std::future::ResumeTy, &'r nats::asynk::client::Client, &'s str, std::option::Option<&'t0 str>, impl std::future::Future, (), nats::asynk::client::Client, &'t1 async_mutex::Mutex<nats::asynk::client::State>, std::sync::Arc<async_mutex::Mutex<nats::asynk::client::State>>, impl std::future::Future, async_mutex::MutexGuard<'t4, nats::asynk::client::State>, &'t5 mut async_mutex::MutexGuard<'t6, nats::asynk::client::State>, impl std::future::Future, u64, nats::asynk::client::State, &'t10 mut std::option::Option<std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't11)>>>, std::option::Option<std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't12)>>>, std::option::Option<&'t13 mut std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't14)>>>, &'t15 mut std::pin::Pin<std::boxed::Box<(dyn futures_io::if_std::AsyncWrite + std::marker::Send + 't16)>>, nats::asynk::proto::ClientOp<'t17>, impl std::future::Future}]>`
   = note: required because it appears within the type `impl std::future::Future`
   = note: required because it appears within the type `impl std::future::Future`
   = note: required because it appears within the type `for<'r, 's, 't0, 't1, 't2, 't3, 't4> {std::future::ResumeTy, &'r nats::asynk::connection::Connection, &'s str, std::option::Option<&'t0 str>, nats::asynk::connection::Connection, &'t1 nats::asynk::client::Client, nats::asynk::client::Client, impl std::future::Future, ()}`
   = note: required because it appears within the type `[static generator@nats::asynk::connection::Connection::do_subscribe::{{closure}}#0 0:&nats::asynk::connection::Connection, 1:&str, 2:std::option::Option<&str> for<'r, 's, 't0, 't1, 't2, 't3, 't4> {std::future::ResumeTy, &'r nats::asynk::connection::Connection, &'s str, std::option::Option<&'t0 str>, nats::asynk::connection::Connection, &'t1 nats::asynk::client::Client, nats::asynk::client::Client, impl std::future::Future, ()}]`
   = note: required because it appears within the type `std::future::from_generator::GenFuture<[static generator@nats::asynk::connection::Connection::do_subscribe::{{closure}}#0 0:&nats::asynk::connection::Connection, 1:&str, 2:std::option::Option<&str> for<'r, 's, 't0, 't1, 't2, 't3, 't4> {std::future::ResumeTy, &'r nats::asynk::connection::Connection, &'s str, std::option::Option<&'t0 str>, nats::asynk::connection::Connection, &'t1 nats::asynk::client::Client, nats::asynk::client::Client, impl std::future::Future, ()}]>`
   = note: required because it appears within the type `impl std::future::Future`
   = note: required because it appears within the type `impl std::future::Future`
   = note: required because it appears within the type `for<'r, 's, 't0, 't1, 't2, 't3> {std::future::ResumeTy, &'r nats::asynk::connection::Connection, &'s str, std::option::Option<&'t0 str>, impl std::future::Future, ()}`
   = note: required because it appears within the type `[static generator@nats::asynk::connection::Connection::subscribe::{{closure}}#0 0:&nats::asynk::connection::Connection, 1:&str for<'r, 's, 't0, 't1, 't2, 't3> {std::future::ResumeTy, &'r nats::asynk::connection::Connection, &'s str, std::option::Option<&'t0 str>, impl std::future::Future, ()}]`
   = note: required because it appears within the type `std::future::from_generator::GenFuture<[static generator@nats::asynk::connection::Connection::subscribe::{{closure}}#0 0:&nats::asynk::connection::Connection, 1:&str for<'r, 's, 't0, 't1, 't2, 't3> {std::future::ResumeTy, &'r nats::asynk::connection::Connection, &'s str, std::option::Option<&'t0 str>, impl std::future::Future, ()}]>`
   = note: required because it appears within the type `impl std::future::Future`
   = note: required because it appears within the type `impl std::future::Future`
   = note: required because it appears within the type `for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10> {std::future::ResumeTy, &'r nats::asynk::connection::Connection, &'s str, &[u8], std::string::String, &'t0 std::string::String, impl std::future::Future, (), nats::asynk::subscription::Subscription, std::option::Option<&'t3 str>, std::option::Option<&'t4 nats::headers::Headers>, impl std::future::Future, &'t9 mut nats::asynk::subscription::Subscription, futures_lite::stream::NextFuture<'t10, nats::asynk::subscription::Subscription>}`
   = note: required because it appears within the type `[static generator@nats::asynk::connection::Connection::request::{{closure}}#0 0:&nats::asynk::connection::Connection, 1:&str, 2:&[u8] for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10> {std::future::ResumeTy, &'r nats::asynk::connection::Connection, &'s str, &[u8], std::string::String, &'t0 std::string::String, impl std::future::Future, (), nats::asynk::subscription::Subscription, std::option::Option<&'t3 str>, std::option::Option<&'t4 nats::headers::Headers>, impl std::future::Future, &'t9 mut nats::asynk::subscription::Subscription, futures_lite::stream::NextFuture<'t10, nats::asynk::subscription::Subscription>}]`
   = note: required because it appears within the type `std::future::from_generator::GenFuture<[static generator@nats::asynk::connection::Connection::request::{{closure}}#0 0:&nats::asynk::connection::Connection, 1:&str, 2:&[u8] for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10> {std::future::ResumeTy, &'r nats::asynk::connection::Connection, &'s str, &[u8], std::string::String, &'t0 std::string::String, impl std::future::Future, (), nats::asynk::subscription::Subscription, std::option::Option<&'t3 str>, std::option::Option<&'t4 nats::headers::Headers>, impl std::future::Future, &'t9 mut nats::asynk::subscription::Subscription, futures_lite::stream::NextFuture<'t10, nats::asynk::subscription::Subscription>}]>`
   = note: required because it appears within the type `impl std::future::Future`
   = note: required because it appears within the type `impl std::future::Future`
   = note: required because it appears within the type `for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6> {std::future::ResumeTy, &'r ServiceImpl, &'s Data, ServiceImpl, &'t0 nats::asynk::connection::Connection, nats::asynk::connection::Connection, &'t1 str, Data, &'t2 std::string::String, std::string::String, &'t3 [u8], impl std::future::Future, ()}`
   = note: required because it appears within the type `[static generator@src/main.rs:33:49: 36:6 _self:&ServiceImpl, data:&Data for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6> {std::future::ResumeTy, &'r ServiceImpl, &'s Data, ServiceImpl, &'t0 nats::asynk::connection::Connection, nats::asynk::connection::Connection, &'t1 str, Data, &'t2 std::string::String, std::string::String, &'t3 [u8], impl std::future::Future, ()}]`
   = note: required because it appears within the type `std::future::from_generator::GenFuture<[static generator@src/main.rs:33:49: 36:6 _self:&ServiceImpl, data:&Data for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6> {std::future::ResumeTy, &'r ServiceImpl, &'s Data, ServiceImpl, &'t0 nats::asynk::connection::Connection, nats::asynk::connection::Connection, &'t1 str, Data, &'t2 std::string::String, std::string::String, &'t3 [u8], impl std::future::Future, ()}]>`
   = note: required because it appears within the type `impl std::future::Future`
   = note: required because it appears within the type `impl std::future::Future`
   = note: required for the cast to the object type `dyn std::future::Future<Output = std::vec::Vec<u8>> + std::marker::Send`
```

**Comment by sergeyzenchenko** [comment]
I've made PR with a fix https://github.com/nats-io/nats.rs/pull/94

**Comment by sergeyzenchenko** [comment]
Can you please take a look at it @stjepang @spacejam ?

**Comment by andrewbanchich** [comment]
I'm using `async_trait` as well so that would explain it!

**Comment by sergeyzenchenko** [comment]
Thanks for merging @stjepang  can you please release a new version with this bugfix? @andrewbanchich I think you can close this issue

**Comment by stjepang** [comment]
@sergeyzenchenko Published v0.7.4
